### PR TITLE
feat(core): [Payment Link] add dynamic merchant fields

### PR DESCRIPTION
**Pull request by Sakilmostak**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

Merchant can send fields in the request which can be shown in the payment link during its rendering. The values should be key value pair and atmost 5 fields can be displayed dynamically.

### Additional Changes

- [x] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## UI Representation of Payment Link:
Without card expanded:
image
With cart expanded:
image
Mobile View:
image
Tablet View:
image

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

File: crates/router/src/core/payment_link.rs [hunk 1]
```diff
@@ -575,7 +590,8 @@ pub fn get_payment_link_config_based_on_priority(
         (
             enabled_saved_payment_method,
             DEFAULT_ENABLE_SAVED_PAYMENT_METHOD
-        )
+        ),
+        (merchant_details, IndexMap::new())
```
Comment (sahkal) [inline crates/router/src/core/payment_link.rs:594]: Why are we sending `IndexMap::new()` i suppose we have defined a const for it in the consts file.

**Review by sahkal — COMMENTED** [review]
Can you please add a video or screenshot depicting your changes, since this has major FE play.

File: crates/router/src/core/payment_link.rs [hunk 2]
```diff
@@ -586,6 +602,11 @@ pub fn get_payment_link_config_based_on_priority(
         display_sdk_only,
         enabled_saved_payment_method,
         allowed_domains,
+        merchant_details: serde_json::to_string(&merchant_details)
+            .change_context(errors::ApiErrorResponse::InvalidDataValue {
+                field_name: "merchant_details",
+            })
+            .ok(),
     };
```
Comment (sahkal) [inline crates/router/src/core/payment_link.rs:610]: What is scope of this field is this transactional data, if yes then it should only be in paymentscreate payment link config. And if that is the case then we donot need this field to go through priority macro

File: crates/router/src/core/payment_link.rs [hunk 3]
```diff
@@ -586,6 +592,18 @@ pub fn get_payment_link_config_based_on_priority(
...
+            payment_link_config
+                .theme_config
+                .merchant_details
+                .and_then(|merchant_details| {
+                    serde_json::to_string(&merchant_details)
+                        .change_context(errors::ApiErrorResponse::InvalidDataValue {
+                            field_name: "merchant_details",
+                        })
+                        .ok()
+                })
```
Comment (sahkal) [inline crates/router/src/core/payment_link.rs:605]: can we also, add a error log for it!

**Review by sahkal — APPROVED** [review]

**Review by sahkal — DISMISSED** [review]

**Review by sahkal — DISMISSED** [review]

File: crates/router/src/core/payment_link.rs [hunk 4]
```diff
@@ -600,6 +606,24 @@ pub fn get_payment_link_config_based_on_priority(
...
+                .theme_config
+                .transaction_details
+                .and_then(|transaction_details| {
+                    match serde_json::to_string(&transaction_details).change_context(
+                        errors::ApiErrorResponse::InvalidDataValue {
+                            field_name: "transaction_details",
+                        },
+                    ) {
+                        Ok(details) => Some(details),
+                        Err(_) => {
```
Comment (jarnura) [inline crates/router/src/core/payment_link.rs:620]: Don't skip the errors, log the error

**Review by jarnura — CHANGES_REQUESTED** [review]

**Review by jarnura — APPROVED** [review]

File: crates/router/src/types/transformers.rs [hunk 5]
```diff
@@ -1740,6 +1740,7 @@ impl ForeignFrom<diesel_models::business_profile::PaymentLinkConfigRequest>
             sdk_layout: item.sdk_layout,
             display_sdk_only: item.display_sdk_only,
             enabled_saved_payment_method: item.enabled_saved_payment_method,
+            transaction_details: None,
```
[the remark this answers is not in the document]
Comment (Sakilmostak) [reply crates/router/src/types/transformers.rs:1743]: transaction_details will only be picked from payments create. It won't be coming from business profile level thus it is kept as none here

**Review by SanchithHegde — APPROVED** [review]

**Pull request merged**